shared/src/bad_token/trace_call.rs

```diff
@@ -16,6 +16,31 @@ use web3::{
     types::{BlockTrace, CallRequest, Res},
 };
 
+/// To detect bad tokens we need to find some address on the network that owns the token so that we
+/// can use it in our simulations.
+#[async_trait::async_trait]
+pub trait TokenOwnerFinding: Send + Sync {
+    /// Find candidate addresses that might own the token.
+    async fn find_candidate_owners(&self, token: H160) -> Result<Vec<H160>>;
+}
+
+pub struct AmmPairProviderFinder {
+    pub inner: Arc<dyn AmmPairProvider>,
+    pub base_tokens: Vec<H160>,
+}
+
+#[async_trait::async_trait]
+impl TokenOwnerFinding for AmmPairProviderFinder {
+    async fn find_candidate_owners(&self, token: H160) -> Result<Vec<H160>> {
+        Ok(self
+            .base_tokens
+            .iter()
+            .filter_map(|&base_token| TokenPair::new(base_token, token))
+            .map(|pair| self.inner.pair_address(&pair))
+            .collect())
```

Comment:
I am not sure how this is finding potential owners of a token.... Oh! are we enumerating the pool addresses containing that token? Clever.